src/Neo.VM/ReferenceCounter.cs

```diff
@@ -22,86 +22,180 @@ namespace Neo.VM
+        private LinkedList<HashSet<StackItem>>? _cachedComponents;
+
+        // Keeps the total count of references.
+        private int _referencesCount = 0;
 
         /// <summary>
-        /// Indicates the number of this counter.
+        /// Gets the count of references.
         /// </summary>
-        public int Count => references_count;
+        public int Count => _referencesCount;
 
+        /// <summary>
+        /// Determines if an item needs to be tracked based on its type.
+        /// </summary>
+        /// <param name="item">The item to check.</param>
+        /// <returns>True if the item needs to be tracked, otherwise false.</returns>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
         private static bool NeedTrack(StackItem item)
         {
+            // Track all items if TrackAllItems is true.
 #pragma warning disable CS0162
             if (TrackAllItems) return true;
 #pragma warning restore CS0162
+
+            // Track the item if it is a CompoundType or Buffer.
             if (item is CompoundType or Buffer) return true;
             return false;
         }
 
+        /// <summary>
+        /// Adds a reference to a specified item with a parent compound type.
+        ///
+        /// This method is used when an item gains a new reference through a parent compound type.
+        /// It increments the reference count and updates the tracking structures if necessary.
+        ///
+        /// Use this method when you need to add a reference from a compound type to a stack item.
+        /// </summary>
+        /// <param name="item">The item to add a reference to.</param>
+        /// <param name="parent">The parent compound type.</param>
         internal void AddReference(StackItem item, CompoundType parent)
         {
-            references_count++;
+            // Increment the reference count.
+            _referencesCount++;
+
+            // If the item doesn't need to be tracked, return early.
+            // Only track CompoundType and Buffer items.
             if (!NeedTrack(item)) return;
-            cached_components = null;
-            tracked_items.Add(item);
-            item.ObjectReferences ??= new(ReferenceEqualityComparer.Instance);
+
+            // Invalidate the cached components since the tracked items are changing.
+            _cachedComponents = null;
+
+            // Add the item to the set of tracked items.
+            _trackedItems.Add(item);
+
+            // Initialize the ObjectReferences dictionary if it is null.
+            item.ObjectReferences ??= new Dictionary<CompoundType, StackItem.ObjectReferenceEntry>(ReferenceEqualityComparer.Instance);
+
+            // Add the parent to the item's ObjectReferences dictionary and increment its reference count.
             if (!item.ObjectReferences.TryGetValue(parent, out var pEntry))
             {
-                pEntry = new(parent);
+                pEntry = new StackItem.ObjectReferenceEntry(parent);
```

Comment:
Can you revert this change from here?

Reply:
Problems like this are very minor, not worth a pr to fix, but if we don't fix it, it will always be this unclear syntax.

Reply:
Fix in another PR, it is better.
Let just comments here

Even the underscore could be moved to another PR.